[BUG] Can't load external tools due to missing schemas

## Bug Description

When using tools built following the instructions everything works as expected, but when I want to use an NPM tool from a third party the schemas are not being parsed correctly.

## Reproduction Steps

Import any external server. IE: 
https://www.npmjs.com/package/time-mcp

IE:
```
// mcp_bug_report.js
// Minimal example to demonstrate potential schema loading issue with @langchain/mcp-adapters

import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import { loadMcpTools } from '@langchain/mcp-adapters';

async function demonstrateSchemaIssue() {
  // Configuration based on mcp.json for 'time-mcp'
  const serverName = 'time-mcp';
  const command = 'npx';
  const args = ['-y', 'time-mcp'];

  console.log(`[INFO] Setting up transport for ${serverName}...`);
  const transport = new StdioClientTransport({ command, args });

  console.log(`[INFO] Initializing base MCP client for ${serverName}...`);
  const client = new Client({
    name: 'mcp-bug-report-client',
    version: '1.0.0',
  });

  try {
    console.log(`[INFO] Connecting client to transport for ${serverName}...`);
    await client.connect(transport);
    console.log(`[INFO] Successfully connected to ${serverName}.`);

    console.log(`[INFO] Loading tools from ${serverName} using loadMcpTools...`);
    // Load tools using the adapter function
    const tools = await loadMcpTools(serverName, client);
    console.log(`[INFO] ${tools.length} tools loaded.`);

    console.log('\n[DEBUG] Inspecting loaded tools:');
    if (tools.length > 0) {
      tools.forEach((tool, index) => {
        console.log(`\n--- Tool ${index + 1} ---`);
        console.log(`Name: ${tool.name}`);
        console.log(`Description: ${tool.description}`);
        // The core issue: Inspect the 'schema' property
        console.log(`Schema: ${JSON.stringify(tool.schema, null, 2)}`);
        // Also log the entire tool object for context
        // console.log(`Full Tool Object: ${JSON.stringify(tool, null, 2)}`);
      });
      console.log('\n[EXPECTATION] The "Schema" output above should show a valid Zod schema object defining the tool\'s parameters.');
      console.log('[OBSERVATION] The "Schema" output might be undefined, null, or not a proper Zod schema.');
    } else {
      console.log('[WARN] No tools were found on the server.');
    }

  } catch (error) {
    console.error('\n[ERROR] An error occurred during the test:', error);
  } finally {
    if (client) {
      console.log('\n[INFO] Closing client connection...');
      await client.close();
      console.log('[INFO] Connection closed.');
    }
  }
}

demonstrateSchemaIssue();

```

## Expected Behavior

Schemas to be loaded properly like when other tools load the servers IE:
See cline example:

![Image](https://github.com/user-attachments/assets/30bcba37-da1c-4901-82b4-fdb1f00a3370)

## Actual Behavior

Schemas are broken and LLMs don't know how to call the tools

```
[DEBUG] Inspecting loaded tools:

--- Tool 1 ---
Name: current_time
Description: Get the current date and time.
Schema: {
  "_def": {
    "unknownKeys": "strict",
    "catchall": {
      "_def": {
        "typeName": "ZodNever"
      },
      "~standard": {
        "version": 1,
        "vendor": "zod"
      }
    },
    "typeName": "ZodObject"
  },
  "~standard": {
    "version": 1,
    "vendor": "zod"
  },
  "_cached": null
}

--- Tool 2 ---
Name: relative_time
Description: Get the relative time from now.
Schema: {
  "_def": {
    "unknownKeys": "strict",
    "catchall": {
      "_def": {
        "typeName": "ZodNever"
      },
      "~standard": {
        "version": 1,
        "vendor": "zod"
      }
    },
    "typeName": "ZodObject"
  },
  "~standard": {
    "version": 1,
    "vendor": "zod"
  },
  "_cached": null
}

--- Tool 3 ---
Name: days_in_month
Description: Get the number of days in a month. If no date is provided, get the number of days in the current month.
Schema: {
  "_def": {
    "unknownKeys": "strict",
    "catchall": {
      "_def": {
        "typeName": "ZodNever"
      },
      "~standard": {
        "version": 1,
        "vendor": "zod"
      }
    },
    "typeName": "ZodObject"
  },
  "~standard": {
    "version": 1,
    "vendor": "zod"
  },
  "_cached": null
}

--- Tool 4 ---
Name: get_timestamp
Description: Get the timestamp for the time.
Schema: {
  "_def": {
    "unknownKeys": "strict",
    "catchall": {
      "_def": {
        "typeName": "ZodNever"
      },
      "~standard": {
        "version": 1,
        "vendor": "zod"
      }
    },
    "typeName": "ZodObject"
  },
  "~standard": {
    "version": 1,
    "vendor": "zod"
  },
  "_cached": null
}
...

```
## Environment

- OS: Ubuntu 22.04
- Node.js version: v22.14.0
- Package version: 0.3.4

## Additional Context

Add any other context about the problem here, such as:

- Server implementation details: https://www.npmjs.com/package/time-mcp
- Transport type (stdio, SSE): stdio

## Possible Solution

Review how the tools are loaded and how the schemas are parsed. I'll give it a go if someone can confirm

--- a/__tests__/tools.test.ts
+++ b/__tests__/tools.test.ts
@@ -1,6 +1,9 @@
 import { describe, test, expect, beforeEach, vi, MockedObject } from "vitest";
 import type { Client } from "@modelcontextprotocol/sdk/client/index.js";
-import type { StructuredTool } from "@langchain/core/tools";
+import {
+  StructuredTool,
+  ToolInputParsingException,
+} from "@langchain/core/tools";
 import type {
   EmbeddedResource,
   ImageContent,
@@ -57,6 +60,74 @@ describe("Simplified Tool Adapter Tests", () => {
       expect(tools.length).toBe(2);
       expect(tools[0].name).toBe("tool1");
       expect(tools[1].name).toBe("tool2");
+    });
+
+    test("should validate tool input against input schema", async () => {
+      // Set up mock response
+      mockClient.listTools.mockReturnValueOnce(
+        Promise.resolve({
+          tools: [
+            {
+              name: "weather",
+              description: "Get the weather for a given city",
+              inputSchema: {
+                type: "object",
+                properties: {
+                  city: { type: "string" },
+                },
+                required: ["city"],
+              },
+            },
+          ],
+        })
+      );
+
+      mockClient.callTool.mockImplementation((params) => {
+        // should not be called if input is invalid
+        const args = params.arguments as { city: string };
+        expect(args.city).toBeDefined();
+        expect(typeof args.city).toBe("string");
+
+        return Promise.resolve({
+          content: [
+            {
+              type: "text",
+              text: `It is currently 70 degrees and cloudy in ${args.city}.`,
+            },
+          ],
+        });
+      });
+
+      // Load tools
+      const tools = await loadMcpTools(
+        "mockServer(should validate tool input against input schema)",
+        mockClient as Client
+      );
+
+      // Verify results
+      expect(tools.length).toBe(1);
+      expect(tools[0].name).toBe("weather");
+
+      const weatherTool = tools[0];
+
+      // should not invoke the tool when input is invalid
+      await expect(
+        weatherTool.invoke({ location: "New York" })
+      ).rejects.toThrow(ToolInputParsingException);
+
+      expect(mockClient.callTool).not.toHaveBeenCalled();
+
+      // should invoke the tool when input is valid
+      await expect(weatherTool.invoke({ city: "New York" })).resolves.toEqual(
+        "It is currently 70 degrees and cloudy in New York."
+      );
+
+      expect(mockClient.callTool).toHaveBeenCalledWith({
+        arguments: {
+          city: "New York",
+        },
+        name: "weather",
+      });
     });
 
     test("should handle empty tool list", async () => {

--- a/examples/langgraph_example.ts
+++ b/examples/langgraph_example.ts
@@ -10,7 +10,7 @@
  * - Built-in persistence capabilities
  *
  * In this example, we:
- * 1. Set up an MCP client to connect to the MCP Math server reference example
+ * 1. Set up an MCP client to connect to the MCP everything server reference example
  * 2. Create a LangGraph workflow with two nodes: one for the LLM and one for tools
  * 3. Define the edges and conditional routing between the nodes
  * 4. Execute the workflow with example queries
@@ -42,7 +42,7 @@ dotenv.config();
 
 /**
  * Example demonstrating how to use MCP tools with LangGraph agent flows
- * This example connects to a math server and uses its tools
+ * This example connects to a everything server and uses its tools
  */
 async function runExample() {
   let client: MultiServerMCPClient | null = null;
@@ -50,12 +50,12 @@ async function runExample() {
   try {
     console.log("Initializing MCP client...");
 
-    // Create a client with configurations for the math server only
+    // Create a client with configurations for the everything server only
     client = new MultiServerMCPClient({
-      math: {
+      everything: {
         transport: "stdio",
         command: "npx",
-        args: ["-y", "@modelcontextprotocol/server-math"],
+        args: ["-y", "@modelcontextprotocol/server-everything"],
       },
     });
 
@@ -137,7 +137,9 @@ async function runExample() {
     const app = workflow.compile();
 
     // Define queries for testing
-    const queries = ["What is 5 + 3?", "What is 7 * 9?"];
+    const queries = [
+      "If Sally has 420324 apples and mark steals 7824 of them, how many does she have left?",
+    ];
 
     // Test the LangGraph agent with the queries
     console.log("\n=== RUNNING LANGGRAPH AGENT ===");

--- a/package.json
+++ b/package.json
@@ -52,16 +52,16 @@
     "zod": "^3.24.2"
   },
   "peerDependencies": {
-    "@langchain/core": "^0.3.40"
+    "@langchain/core": "^0.3.44"
   },
   "optionalDependencies": {
     "extended-eventsource": "^1.x"
   },
   "devDependencies": {
     "@eslint/js": "^9.21.0",
-    "@langchain/core": "^0.3.43",
+    "@langchain/core": "^0.3.44",
     "@langchain/langgraph": "^0.2.62",
-    "@langchain/openai": "^0.5.2",
+    "@langchain/openai": "^0.5.5",
     "@langchain/scripts": "^0.1.3",
     "@tsconfig/recommended": "^1.0.8",
     "@types/debug": "^4.1.12",
@@ -91,7 +91,8 @@
     "vitest": "^3.1.1"
   },
   "resolutions": {
-    "typescript": "4.9.5"
+    "typescript": "4.9.5",
+    "uuid": "^11.0.0"
   },
   "engines": {
     "node": ">=18"

--- a/src/tools.ts
+++ b/src/tools.ts
@@ -18,18 +18,7 @@ import {
   MessageContentImageUrl,
   MessageContentText,
 } from "@langchain/core/messages";
-import type { JSONSchema } from "@dmitryrechkin/json-schema-to-zod";
-import type { ZodSchema } from "zod";
 import debug from "debug";
-
-let JSONSchemaToZod: { convert: (schema: JSONSchema) => ZodSchema } | undefined;
-
-async function convertSchema(schema: JSONSchema): Promise<ZodSchema> {
-  if (!JSONSchemaToZod) {
-    ({ JSONSchemaToZod } = await import("@dmitryrechkin/json-schema-to-zod"));
-  }
-  return JSONSchemaToZod.convert(schema);
-}
 
 // Replace direct initialization with lazy initialization
 let debugLog: debug.Debugger;
@@ -279,12 +268,7 @@ export async function loadMcpTools(
             const dst = new DynamicStructuredTool({
               name: `${toolNamePrefix}${tool.name}`,
               description: tool.description || "",
-              schema: await convertSchema(
-                (tool.inputSchema ?? {
-                  type: "object",
-                  properties: {},
-                }) as JSONSchema
-              ),
+              schema: tool.inputSchema,
               responseFormat: "content_and_artifact",
               func: _callTool.bind(
                 null,

